### PR TITLE
MSFT: 7429352 JSRT context leak regression fix

### PR DESCRIPTION
**Pull request by Yongqu**
We start supporting cross-context scenarios in JSRT in chakracore. However we are having
leaks in different scenarios due to pinned object. This change cleans up some basic leak scenarios.
We might need to have more cleanup in the future to make sure that there is only one
root for a JSRTContext or script context.

One major leak source is the guest arena in ScriptContext. objects allocated in the guestArena are
hard pinned till the scriptcontext is closed. We need to move objects off the guestArena:
    scriptContext::Cache. it's allocated using normal RecyclerNew, linked in the JavascriptLibrary, and
keep the reference in scriptcontext to minimize the code change.
    BuiltInFunctionMap. We might need to have a full cleanup of BindReference, but for now we just move
out the builtInFunctionMap to JavascripLibrary.
    Convert ScriptContext::BindReference to use linklist in JavascriptLibrary
    Allocate JSRTContext using RecyclerNew (no more leaf); have JSRTContext have a different reference to
JavascriptLibrary to keep it alive, and masked out other fields for potential false reference.

### CODE REVIEW

**Comment by Yongqu** [comment]
@digitalinfinity @kunalspathak can you take a look?

File: lib/Parser/RegexPattern.cpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ namespace UnifiedRegex
             Assert(
                 !GetScriptContext()->GetDynamicRegexMap()->TryGetValue(
                     RegexKey(source.GetBuffer(), source.GetLength(), GetFlags()),
-                    &p) ||
+                    &p) || ( source.GetLength() == 0 ) ||
```
Comment (kunalspathak) [inline lib/Parser/RegexPattern.cpp]: what is this about?

Comment (Yongqu) [inline lib/Parser/RegexPattern.cpp]: We have a builtin regexp stored in the map that are not cleaned up by default. It's ok for it to be there.

**Comment by Yongqu** [comment]
@dotnet-bot please test

**Comment by Yongqu** [comment]
@dotnet-bot test this

**Comment by Yongqu** [comment]
@dotnet-bot test this please

**Comment by digitalinfinity** [comment]
:shipit:

**Pull request merged**